### PR TITLE
Add a skeleton of the 3.5.0 release post.

### PR DESCRIPTION
**Pull request by samphippen**

### CODE REVIEW

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
+---
```
Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: The file name should have `3-5`, not `3.5` -- at least if we want to follow the pattern for all the other release posts.

Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: In fact, I had to rename the file to get middleman to serve this blog post locally.  I guess it requires the dash form.

That suggests that you wrote this without seeing how it renders in middleman.  That's fine, but be sure to check how it renders in your browser.  I've seen plenty of cases where things did not render how you'd expect initially, and the manual check is important before we publish.

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+The headline here is that RSpec 3.5.0 is compatible with Rails 5. As Rails 5
+betas and release candidates have been released, we've been releasing betas of
+3.5.0 to keep up alongside Rails. Due to this being a major release of Rails,
+but not a major release of RSpec, Rails has removed some APIs that we consume.
+This only gets exposed to you, our users, in one place: controller testing.
+
+In [Rails 5](https://github.com/rails/rails/issues/18950) `assigns` and
+`assert_template` are "soft deprecated". Controller tests themselves *are not*,
+and adding `:type => :controller` to your specs is still 100% supported.
+Through Rails 3 and 4: it was both prevalent and idiomatic to use `assigns` in
```
Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: > Through Rails 3 and 4: it was both prevalent and idiomatic

The `:` doesn't make sense to me here.  Why not this instead?

> Through Rails 3 and 4, it was both prevalent and idiomatic

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+controller specs. As this is a minor release of RSpec our commitment to SemVer
+means that we are not going to break your existing controller specs. For
+existing Rails applications that make heavy use of `assigns` adding the
+[`rails-controller-testing`](https://github.com/rails/rails-controller-testing)
+to one's Gemfile will restore `assigns` and `assert_template`. RSpec integrates
```
Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: While I appreciate the gender-neutral "one's" here, I think it reads kinda awkwardly and it reads better IMO to rephrase the sentence to use "your" (and is just as gender-neutral):

> If your existing Rails applications makes heavy use of `assigns`, just add the
> [`rails-controller-testing`](https://github.com/rails/rails-controller-testing) gem
> to your Gemfile to restore `assigns` and `assert_template`.

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+with this gem seamlessly, so your controller specs should just continue to work.
+
+For new Rails apps: we don't recommend adding the `rails-controller-testing gem
```
Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: Missing a closing backtick here.

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+to your application. The "official" stance of the Rails team and the RSpec core
```
Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: s/stance/recommendation/ ?

I think I like the "friendlier" tone of "recommendation" better than "stance".

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+team is that [request specs](https://www.relishapp.com/rspec/rspec-rails/docs/request-specs/request-spec)
+should now be written instead. Request specs allow one to focus on a single
```
Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: "should" kinda has a scolding tone to it, as well.  Maybe that's just my baggage, but I think we can rephrase this to be a bit less emphatic while still getting the point across:

> The official recommendation of the Rails and RSpec core teams is to write [request specs](https://www.relishapp.com/rspec/rspec-rails/docs/request-specs/request-spec) instead.

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+controller action, but unlike controller tests involve the router, the
+middleware stack, and both rack requests and responses. In Rails 5, they are
```
Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: I think it's helpful to state _why_ it matters that request specs involve all these extra things:

> Request specs allow you to focus on a single controller action, but unlike controller
> tests involve the router, the middleware stack, and both rack requests and responses -- which
> avoids a whole host of gotchas and caveats that plagued controller specs.

I think that makes the concrete benefit more obvious.

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+very fast thanks to the work by [Eileen Uchitelle](https://twitter.com/eileencodes?lang=en-gb)
```
Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: Maybe this is just me being anal, but I don't think rails 5 request specs will feel  "very fast" to someone used to writing isolated unit specs.

Perhaps "they are much faster thanks to..." will get the point across just as well?

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+<sup>[[1]](#footnotes)</a></sup> of the Rails Committer Team.
```
Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: The `</a>` here isn't needed and just causes the tag literal to be rendered.

BTW, you don't need to do all this `<sup>[[1]](#footnotes)</sup>` business.

See https://raw.githubusercontent.com/rspec/rspec.github.io/source/source/blog/2012-06-15-rspecs-new-expectation-syntax.md for examples of how we've done footnotes before. You just need to do `[^foot_1]`.

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+The other important feature of Rails 5 we wanted to discuss is ActionCable.
+Unfortunately RSpec is not able to provide a simple testing story for
+ActionCable at this time. Rails is working on a testing type for ActionCable
+slated for release as part of Rails 5.1. We'll be watching that closely and work
+something up when it's ready. In the mean time, we suggest you test ActionCable
+through a browser, in an integrated fashion.
```
Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: Rails 5 has not support for automated tests against ActionCable?  Wow!  Weird.

Anyhow, instead of telling readers to just test manually (that seems awful), maybe we can tell them they'll need to test it in integrated fashion using a real web socket client such as https://github.com/shokai/websocket-client-simple ?  After all, there's no reason they couldn't write tests that boot rails and use a client like that to interact with it, right?

**Comment by myronmarston** [comment]
It's a great start, @samphippen.  Nice work!

Let me know when you've rename the file and then I can get started filling in some of the notable changes.

@yujinakayama -- are you planning to do a Japanese translation again?

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md [hunk 11]
```diff
@@ -0,0 +1,125 @@
+---
+title: RSpec 3.5 has been released!
+author: Sam Phippen
+---
+
+RSpec 3.5 has just been released! Given our commitment to
+[semantic versioning](http://semver.org/), this should be a trivial
+upgrade for anyone already using any RSpec 3 release, but if we did introduce
```
Comment (JonRowe) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: s/any/an/ any reads funny to me

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+Thank you to everyone who helped make this release happen!
+
+## Notable Changes
+
+### Rails: Support for Rails 5
+
+The headline here is that RSpec 3.5.0 is compatible with Rails 5. As Rails 5
+betas and release candidates have been released, we've been releasing betas of
+3.5.0 to keep up alongside Rails. Due to this being a major release of Rails,
+but not a major release of RSpec, Rails has removed some APIs that we consume.
```
Comment (JonRowe) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: Rails removing APIs is not due to to RSpec ;)

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md (hunk 8 above)
Comment (JonRowe) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: or just as fast as controller tests, has anyone actually benchmarked the difference?

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+to your application. The "official" stance of the Rails team and the RSpec core
+team is that [request specs](https://www.relishapp.com/rspec/rspec-rails/docs/request-specs/request-spec)
+should now be written instead. Request specs allow one to focus on a single
+controller action, but unlike controller tests involve the router, the
+middleware stack, and both rack requests and responses. In Rails 5, they are
+very fast thanks to the work by [Eileen Uchitelle](https://twitter.com/eileencodes?lang=en-gb)
+<sup>[[1]](#footnotes)</a></sup> of the Rails Committer Team.
+
+The other important feature of Rails 5 we wanted to discuss is ActionCable.
+Unfortunately RSpec is not able to provide a simple testing story for
```
Comment (JonRowe) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: I'm not sure I like "testing story", testing strategy or test interface would be better IMO

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md (hunk 1 above)
Comment (samphippen) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: done

**Comment by samphippen** [comment]
I renamed the file and GitHub collapsed all the feedback. I'll address the rest of it tomorrow. @myronmarston do you want to add any points that you have in the mean time?

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md (hunk 2 above)
Comment (samphippen) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: done

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md (hunk 3 above)
Comment (samphippen) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: done

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md (hunk 4 above)
Comment (samphippen) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: fixed.

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md (hunk 5 above)
Comment (samphippen) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: done.

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md (hunk 6 above)
Comment (samphippen) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: done.

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md (hunk 7 above)
Comment (samphippen) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: done

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md (hunk 9 above)
Comment (samphippen) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: added.

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md (hunk 10 above)
Comment (samphippen) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: part of the problem with this advice is that actioncable has tight integration with rails's own javascript, so I feel like advising our users to test in a browser environment is the best thing for now.

Testing things like notifications show up in the UI is going to be the least coupling increasing way to test action cable right now.  Based on a few conversations I've had with rails maintainers, this seems to be the thing that they want people to do.

What do you think?

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md (hunk 12 above)
Comment (samphippen) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: fixed

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md (hunk 13 above)
Comment (samphippen) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: here story is as in agile, the parlance feels correct to me. Do you think it's fine? How does it read to an american? /cc @myronmarston @cupakromer

**Comment by myronmarston** [comment]
I pushed my first round of updates. Will do more later.

File: source/blog/2016-06-14-rspec-3.5-has-been-released.md (hunk 13 above)
Comment (JonRowe) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: We don't use this parlance anywhere else so it still feels out of context, and whilst story is a process used for developing features it generally isn't used to refer to them after...

Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3.5-has-been-released.md]: I think "story" is ok but maybe we can come up with something everyone is happy with.

**Comment by myronmarston** [comment]
OK, I pushed some more updates.  I've written up everything I had planned.  Anyone want to review?

As far as I know, this is ready to publish besides the stuff we'll fill in or replace on the day of release...except for one thing.  The file name indicates a date of 2016-06-14, which isn't accurate.  @samphippen, feel free to update that now or wait until release day.

File: source/blog/2016-06-14-rspec-3-5-has-been-released.md [hunk 14]
```diff
@@ -0,0 +1,345 @@
+---
+title: RSpec 3.5 has been released!
+author: Sam Phippen
```
Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3-5-has-been-released.md]: Since I wrote up a bunch of the notable changes section, should we change this to `Sam Phippen and Myron Marston`?

Comment (samphippen) [inline source/blog/2016-06-14-rspec-3-5-has-been-released.md]: 👍

Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3-5-has-been-released.md]: Done.

**Comment by JonRowe** [comment]
I've added a snippet about the new respond_to matcher stuff

File: source/blog/2016-06-14-rspec-3-5-has-been-released.md [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+end
+~~~
+
+### Expectations: Keyword argument support for the `respond_to` matcher.
+
+Keyword arguments have been a stable language feature for some time now but RSpec
+has lacked support for placing expectations upon keyword arguments in a variety of our matchers.
+
+In rspec-expectations 3.5 we add support for checking whether an object responds to a method
+using keyword arguments. You'll now be able to check
```
Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3-5-has-been-released.md]: > You'll now be able to check

Seems like this sentence is truncated?

File: source/blog/2016-06-14-rspec-3-5-has-been-released.md [hunk 16]
```diff
@@ -0,0 +1,367 @@
...
+
+~~~ ruby
+  expect(my_object).to respond_to(:find).with_keywords(:limit, :offset) }
+  expect(my_object).to respond_to(:find).with(1).argument.and_keywords(:limit, :offset)
```
Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3-5-has-been-released.md]: No need for the indentation.  It'll help keep it in the width of the blog post if it does not have that.

File: source/blog/2016-06-14-rspec-3-5-has-been-released.md [hunk 17]
```diff
@@ -0,0 +1,367 @@
...
+~~~
+
+ We're also expanding the matchers existing capabilities, adding the ability
+ to check for a range of arguments, or unlimited arguments in addition to a specific
+ number. e.g:
+
+~~~ ruby
+  expect(my_object).to respond_to(:build).with(2..3).arguments
+  expect(my_object).to respond_to(:build).with_unlimited_arguments
+~~~
```
Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3-5-has-been-released.md]: We should acknowledge the contributor who provided this feature.  In general, I try to always thank whoever built the feature.  The exception is myself...since I have usually written the blog posts I always thought it would be weird to thank myself so I leave that out.

File: source/blog/2016-06-14-rspec-3-5-has-been-released.md [hunk 18]
```diff
@@ -0,0 +1,368 @@
...
+end
+~~~
+
+### Expectations: Keyword argument support for the `respond_to` matcher.
+
+Keyword arguments have been a stable language feature for some time now but RSpec
+has lacked support for placing expectations upon keyword arguments in a variety of our matchers.
+
+In rspec-expectations 3.5 we add support for checking whether an object responds to a method
+using keyword arguments. You'll now be able to check a methods response signature for particular
```
Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3-5-has-been-released.md]: s/methods/method's/

File: source/blog/2016-06-14-rspec-3-5-has-been-released.md [hunk 19]
```diff
@@ -0,0 +1,368 @@
...
+keywords as well as arguments.
```
Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3-5-has-been-released.md]: > keywords as well as arguments.

Keyword args are arguments, so this reads a little funny.  How about:

> keywords as well as argument counts.

File: source/blog/2016-06-14-rspec-3-5-has-been-released.md [hunk 20]
```diff
@@ -0,0 +1,368 @@
...
+
+~~~ ruby
+  expect(my_object).to respond_to(:find).with_keywords(:limit, :offset) }
+  expect(my_object).to respond_to(:find).with(1).argument.and_keywords(:limit, :offset)
```
Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3-5-has-been-released.md]: One thing that's unclear to me in this example (and that perhaps warrants changes in rspec-expectations before we are locked into the semantics) is what it means to mix an argument count and keywords.  Is the `(1).argument` bit specifying the number of non-keyword arguments (e.g. positional arguments)?  Where it gets confusing is that at the call site keyword args are just passed as a single hash arg, which is then interpreted as keyword args or as a single positional arg depending on how the method is defined.  So it's not clear to me if keyword args should be counted and if so, if they should be counted as one arg (since the underlying mechanism uses a single hash) or as one arg per keyword.

File: source/blog/2016-06-14-rspec-3-5-has-been-released.md [hunk 21]
```diff
@@ -0,0 +1,370 @@
...
+Keyword arguments have been a stable language feature for some time now but RSpec
+has lacked support for placing expectations upon keyword arguments in a variety of our matchers.
+
+In rspec-expectations 3.5 we add support for checking whether an object responds to a method
+using keyword arguments. You'll now be able to check a methods response signature for particular
+keywords as well as arguments.
+
+~~~ ruby
+expect(my_object).to respond_to(:find).with_keywords(:limit, :offset) }
+expect(my_object).to respond_to(:find).with(1).argument.and_keywords(:limit, :offset)
```
Comment (myronmarston) [inline source/blog/2016-06-14-rspec-3-5-has-been-released.md]: One thing that's unclear to me in this example (and that perhaps warrants changes in rspec-expectations before we are locked into the semantics) is what it means to mix an argument count and keywords.  Is the `(1).argument` bit specifying the number of non-keyword arguments (e.g. positional arguments)?  Where it gets confusing is that at the call site keyword args are just passed as a single hash arg, which is then interpreted as keyword args or as a single positional arg depending on how the method is defined.  So it's not clear to me if keyword args should be counted and if so, if they should be counted as one arg (since the underlying mechanism uses a single hash) or as one arg per keyword.

Comment (JonRowe) [inline source/blog/2016-06-14-rspec-3-5-has-been-released.md]: Argument in this sense means a traditional unnamed argument.

File: source/blog/2016-06-14-rspec-3-5-has-been-released.md (hunk 19 above)
Comment (JonRowe) [inline source/blog/2016-06-14-rspec-3-5-has-been-released.md]: reworded

**Comment by myronmarston** [comment]
Any update on the release date for rails 5, @samphippen?

**Comment by samphippen** [comment]
The release got knocked back, so we don't need to do our release yet.

**Comment by myronmarston** [comment]
> The release got knocked back, so we don't need to do our release yet.

I figured that, since I saw no news about Rails 5 being released.  Is there a new ETA?

**Comment by samphippen** [comment]
I will find out and get back to you.

**Comment by yujinakayama** [comment]
> @yujinakayama -- are you planning to do a Japanese translation again?

Doing now :)
